src/openrct2/Speed.hpp

```diff
@@ -0,0 +1,136 @@
+        using BestValue = std::conditional_t<(sizeof(Value) > sizeof(OtherType)), Value, OtherType>;
+        template<typename OtherType> using BestType = SpeedTemplate<BestValue<OtherType>>;
+
+        template<typename RhsT> constexpr bool operator==(SpeedTemplate<RhsT> const& rhs) const
+        {
+            return BestType<RhsT>(*this).value == BestType<RhsT>(rhs).value;
+        }
+        template<typename RhsT> constexpr bool operator!=(SpeedTemplate<RhsT> const& rhs) const
+        {
+            return !(*this == rhs);
+        }
+        template<typename RhsT> constexpr bool operator>(SpeedTemplate<RhsT> const& rhs) const
+        {
+            return BestType<RhsT>(*this).value > BestType<RhsT>(rhs).value;
+        }
+        template<typename RhsT> constexpr bool operator<=(SpeedTemplate<RhsT> const& rhs) const
+        {
+            return !(*this > rhs);
+        }
+        template<typename RhsT> constexpr bool operator<(SpeedTemplate<RhsT> const& rhs) const
+        {
+            return BestType<RhsT>(*this).value < BestType<RhsT>(rhs).value;
+        }
+        template<typename RhsT> constexpr bool operator>=(SpeedTemplate<RhsT> const& rhs) const
+        {
+            return !(*this < rhs);
+        }
+        constexpr SpeedTemplate<Value>& operator-()
+        {
+            value = -value;
+            return *this;
+        }
+        constexpr SpeedTemplate<Value>& operator+=(SpeedTemplate<Value> const& rhs)
+        {
+            value += rhs.value;
+            return *this;
+        }
+        constexpr SpeedTemplate<Value>& operator-=(SpeedTemplate<Value> const& rhs)
+        {
+            value -= rhs.value;
+            return *this;
+        }
+        template<typename RhsT> constexpr BestType<RhsT> operator+(SpeedTemplate<RhsT> const& rhs) const
+        {
+            return BestType<RhsT>(BestType<RhsT>(*this).value + BestType<RhsT>(rhs).value);
+        }
+        template<typename RhsT> constexpr BestType<RhsT> operator-(SpeedTemplate<RhsT> const& rhs) const
+        {
+            return BestType<RhsT>(BestType<RhsT>(*this).value - BestType<RhsT>(rhs).value);
+        }
+        template<typename RhsT> constexpr BestValue<RhsT> operator/(SpeedTemplate<RhsT> const& rhs) const
+        {
+            return BestType<RhsT>(*this).value / BestType<RhsT>(rhs).value;
+        }
+        constexpr SpeedTemplate<Value> operator/(uint32_t const& rhs) const
+        {
+            return SpeedTemplate<Value>(value / rhs);
+        }
+        constexpr SpeedTemplate<Value> operator*(uint32_t const& rhs) const
+        {
+            return SpeedTemplate<Value>(value * rhs);
+        }
+    };
+
+    using Speed32 = SpeedTemplate<int32_t>;
+    using Speed16 = SpeedTemplate<int16_t>;
+
+    constexpr auto speed16Null = Speed16(-1);
+    // Truncates only use if safe to lose information
+    constexpr Speed16 toSpeed16(Speed32 speed)
+    {
+        return Speed16(speed.getRaw() / 65535);
```

Comment:
Very slightly off error I think, but shouldn't this be 65536?
Pinging @duncanspumpkin since you probably know best.

Reply:
Hmm that's a good point. Not sure how I haven't had issues with that in loco

Reply:
It's only a 0,0015% difference in speed, so hard to notice, but it would be best to be exact. Any `static_assert`s checking for `Speed16` may not be high enough to catch this.